diff --git a/core/modules/datetime_range/config/schema/datetime_range.schema.yml b/core/modules/datetime_range/config/schema/datetime_range.schema.yml
index bb244553ed..d2a6f153a8 100644
--- a/core/modules/datetime_range/config/schema/datetime_range.schema.yml
+++ b/core/modules/datetime_range/config/schema/datetime_range.schema.yml
@@ -6,6 +6,10 @@
 field.storage_settings.daterange:
   type: field.storage_settings.datetime
   label: 'Date range settings'
+  mapping:
+    optional_end_date:
+      type: boolean
+      label: 'Optional end date'
 
 field.value.daterange:
   type: mapping
diff --git a/core/modules/datetime_range/datetime_range.install b/core/modules/datetime_range/datetime_range.install
new file mode 100644
index 0000000000..8f7504c481
--- /dev/null
+++ b/core/modules/datetime_range/datetime_range.install
@@ -0,0 +1,54 @@
+<?php
+
+/**
+ * @file
+ * Install, update and uninstall functions for the Datetime range module.
+ */
+
+/**
+ * Allow NULL in the field end_value column.
+ */
+function datetime_range_update_8900() {
+  $database_schema = \Drupal::database()->schema();
+  $storage_schema = \Drupal::keyValue('entity.storage_schema.sql');
+  $configs = \Drupal::entityTypeManager()
+    ->getStorage('field_storage_config')
+    ->loadByProperties(['type' => 'daterange']);
+
+  foreach ($configs as $config) {
+    $schema_key = $config->getTargetEntityTypeId() . '.field_schema_data.' . $config->getName();
+    $tables = $storage_schema->get($schema_key);
+
+    $field_name = $config->getName() . '_end_value';
+    foreach ($tables as $table_name => &$table) {
+      $table['fields'][$field_name]['not null'] = FALSE;
+      $database_schema->changeField($table_name, $field_name, $field_name, $table['fields'][$field_name]);
+    }
+
+    $storage_schema->set($schema_key, $tables);
+  }
+}
+
+/**
+ * Allow NULL in the field value column, to allow having an end date only.
+ */
+function datetime_range_update_8901() {
+  $database_schema = \Drupal::database()->schema();
+  $storage_schema = \Drupal::keyValue('entity.storage_schema.sql');
+  $configs = \Drupal::entityTypeManager()
+    ->getStorage('field_storage_config')
+    ->loadByProperties(['type' => 'daterange']);
+
+  foreach ($configs as $config) {
+    $schema_key = $config->getTargetEntityTypeId() . '.field_schema_data.' . $config->getName();
+    $tables = $storage_schema->get($schema_key);
+
+    $field_name = $config->getName() . '_value';
+    foreach ($tables as $table_name => &$table) {
+      $table['fields'][$field_name]['not null'] = FALSE;
+      $database_schema->changeField($table_name, $field_name, $field_name, $table['fields'][$field_name]);
+    }
+
+    $storage_schema->set($schema_key, $tables);
+  }
+}
diff --git a/core/modules/datetime_range/src/DateTimeRangeTrait.php b/core/modules/datetime_range/src/DateTimeRangeTrait.php
index 0563f7ff72..240e6927a7 100644
--- a/core/modules/datetime_range/src/DateTimeRangeTrait.php
+++ b/core/modules/datetime_range/src/DateTimeRangeTrait.php
@@ -31,13 +31,13 @@ public function viewElements(FieldItemListInterface $items, $langcode) {
     $separator = $this->getSetting('separator');
 
     foreach ($items as $delta => $item) {
-      if (!empty($item->start_date) && !empty($item->end_date)) {
+      if (!empty($item->start_date)) {
         /** @var \Drupal\Core\Datetime\DrupalDateTime $start_date */
         $start_date = $item->start_date;
         /** @var \Drupal\Core\Datetime\DrupalDateTime $end_date */
         $end_date = $item->end_date;
 
-        if ($start_date->getTimestamp() !== $end_date->getTimestamp()) {
+        if ($end_date !== NULL && $start_date->getTimestamp() !== $end_date->getTimestamp()) {
           $elements[$delta] = $this->renderStartEndWithIsoAttribute($start_date, $separator, $end_date);
         }
         else {
diff --git a/core/modules/datetime_range/src/Plugin/Field/FieldType/DateRangeItem.php b/core/modules/datetime_range/src/Plugin/Field/FieldType/DateRangeItem.php
index 85680e6914..dcc1a0198b 100644
--- a/core/modules/datetime_range/src/Plugin/Field/FieldType/DateRangeItem.php
+++ b/core/modules/datetime_range/src/Plugin/Field/FieldType/DateRangeItem.php
@@ -30,6 +30,15 @@
 )]
 class DateRangeItem extends DateTimeItem {
 
+  /**
+   * {@inheritdoc}
+   */
+  public static function defaultStorageSettings() {
+    return [
+      'optional_end_date' => FALSE,
+    ] + parent::defaultStorageSettings();
+  }
+
   /**
    * Value for the 'datetime_type' setting: store a date and time.
    */
@@ -41,7 +50,7 @@ class DateRangeItem extends DateTimeItem {
   public static function propertyDefinitions(FieldStorageDefinitionInterface $field_definition) {
     $properties['value'] = DataDefinition::create('datetime_iso8601')
       ->setLabel(t('Start date value'))
-      ->setRequired(TRUE);
+      ->setRequired(FALSE);
 
     $properties['start_date'] = DataDefinition::create('any')
       ->setLabel(t('Computed start date'))
@@ -52,7 +61,7 @@ public static function propertyDefinitions(FieldStorageDefinitionInterface $fiel
 
     $properties['end_value'] = DataDefinition::create('datetime_iso8601')
       ->setLabel(t('End date value'))
-      ->setRequired(TRUE);
+      ->setRequired(FALSE);
 
     $properties['end_date'] = DataDefinition::create('any')
       ->setLabel(t('Computed end date'))
@@ -89,6 +98,12 @@ public function storageSettingsForm(array &$form, FormStateInterface $form_state
 
     $element['datetime_type']['#options'][static::DATETIME_TYPE_ALLDAY] = $this->t('All Day');
 
+    $element['optional_end_date'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Optional end date'),
+      '#default_value' => $this->getSetting('optional_end_date'),
+    ];
+
     return $element;
   }
 
@@ -136,4 +151,28 @@ public function onChange($property_name, $notify = TRUE) {
     parent::onChange($property_name, $notify);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function getConstraints() {
+    $constraint_manager = \Drupal::typedDataManager()
+      ->getValidationConstraintManager();
+
+    if (!empty($this->getSetting('optional_end_date'))) {
+      return parent::getConstraints();
+    }
+
+    $label = $this->getFieldDefinition()->getLabel();
+    $constraints[] = $constraint_manager
+      ->create('ComplexData', [
+        'end_value' => [
+          'NotNull' => [
+            'message' => $this->t('The @title end date is required', ['@title' => $label]),
+          ],
+        ],
+      ]);
+
+    return $constraints;
+  }
+
 }
diff --git a/core/modules/datetime_range/src/Plugin/Field/FieldWidget/DateRangeWidgetBase.php b/core/modules/datetime_range/src/Plugin/Field/FieldWidget/DateRangeWidgetBase.php
index b0f3ea7be7..ee894f56da 100644
--- a/core/modules/datetime_range/src/Plugin/Field/FieldWidget/DateRangeWidgetBase.php
+++ b/core/modules/datetime_range/src/Plugin/Field/FieldWidget/DateRangeWidgetBase.php
@@ -19,6 +19,7 @@ class DateRangeWidgetBase extends DateTimeWidgetBase {
    */
   public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
     $element = parent::formElement($items, $delta, $element, $form, $form_state);
+    $optional_end_date = $this->getFieldSetting('optional_end_date');
 
     // Wrap all of the select elements with a fieldset.
     $element['#theme_wrappers'][] = 'fieldset';
@@ -30,6 +31,11 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
       '#title' => $this->t('End date'),
     ] + $element['value'];
 
+    if ($element['#required'] && $optional_end_date) {
+      $element['end_value']['#required'] = FALSE;
+      $element['#required'] = FALSE;
+    }
+
     if ($items[$delta]->start_date) {
       /** @var \Drupal\Core\Datetime\DrupalDateTime $start_date */
       $start_date = $items[$delta]->start_date;
diff --git a/core/modules/datetime_range/tests/src/Functional/EntityResource/EntityTest/EntityTestDateRangeTest.php b/core/modules/datetime_range/tests/src/Functional/EntityResource/EntityTest/EntityTestDateRangeTest.php
index 49017993f9..78515409cf 100644
--- a/core/modules/datetime_range/tests/src/Functional/EntityResource/EntityTest/EntityTestDateRangeTest.php
+++ b/core/modules/datetime_range/tests/src/Functional/EntityResource/EntityTest/EntityTestDateRangeTest.php
@@ -141,7 +141,7 @@ protected function assertNormalizationEdgeCases($method, Url $url, array $reques
       unset($normalization[static::$fieldName][0]['end_value']);
       $request_options[RequestOptions::BODY] = $this->serializer->encode($normalization, static::$format);
       $response = $this->request($method, $url, $request_options);
-      $message = "Unprocessable Entity: validation failed.\n{$fieldName}.0.end_value: This value should not be null.\n";
+      $message = "Unprocessable Entity: validation failed.\n{$fieldName}.0.end_value: The field_daterange end date is required\n";
       $this->assertResourceErrorResponse(422, $message, $response);
 
       // DX: 422 when 'end_value' data type is incorrect.
diff --git a/core/modules/datetime_range/tests/src/Kernel/DateRangeItemTest.php b/core/modules/datetime_range/tests/src/Kernel/DateRangeItemTest.php
index 08aceceab5..24c4d2b561 100644
--- a/core/modules/datetime_range/tests/src/Kernel/DateRangeItemTest.php
+++ b/core/modules/datetime_range/tests/src/Kernel/DateRangeItemTest.php
@@ -105,4 +105,60 @@ public function testDateOnly(): void {
     $this->assertEquals('12:00:00', $end_date->format('H:i:s'));
   }
 
+  /**
+   * Test optional end date.
+   */
+  public function testOptionalEndDate() {
+    $field_name = $this->fieldStorage->getName();
+
+    $this->fieldStorage->setSettings([
+      'datetime_type' => DateRangeItem::DATETIME_TYPE_DATE,
+      'optional_end_date' => FALSE,
+    ])
+      ->save();
+
+    $value = [
+      'value' => '2016-09-21',
+      'end_value' => NULL,
+    ];
+
+    $end_date_without_start_date_value = [
+      'value' => NULL,
+      'end_value' => '2016-09-23',
+    ];
+
+    // Verify entity without optional_end_date enabled.
+    $entity = EntityTest::create([
+      'name' => $this->randomString(),
+      $field_name => $value,
+    ]);
+    $this->assertNotEquals(count($entity->validate()), 0);
+
+    // Verify entity with the optional_end_date enabled.
+    $this->fieldStorage->setSetting('optional_end_date', TRUE)
+      ->save();
+    $entity = EntityTest::create([
+      'name' => $this->randomString(),
+      $field_name => $value,
+    ]);
+    $this->entityValidateAndSave($entity);
+
+    // Verify changing the date value.
+    $this->assertEquals($entity->{$field_name}->end_value, $value['end_value']);
+
+    // Verify entity with the optional_end_date enabled when setting an
+    // end_date but no start date.
+    $this->fieldStorage->setSetting('optional_end_date', TRUE)
+      ->save();
+    $entity = EntityTest::create([
+      'name' => $this->randomString(),
+      $field_name => $end_date_without_start_date_value,
+    ]);
+    $this->entityValidateAndSave($entity);
+
+    // Verify changing the date and end_date value.
+    $this->assertEquals($entity->{$field_name}->end_value, $value['end_value']);
+    $this->assertEquals($entity->{$field_name}->value, $value['value']);
+  }
+
 }
