diff --git a/core/core.services.yml b/core/core.services.yml
index d7a2d63fd06..3e152b78fe6 100644
--- a/core/core.services.yml
+++ b/core/core.services.yml
@@ -496,6 +496,7 @@ services:
   Drupal\Core\Form\FormSubmitterInterface: '@form_submitter'
   form_error_handler:
     class: Drupal\Core\Form\FormErrorHandler
+    arguments: ['@renderer']
   Drupal\Core\Form\FormErrorHandlerInterface: '@form_error_handler'
   form_cache:
     class: Drupal\Core\Form\FormCache
diff --git a/core/lib/Drupal/Core/Form/FormErrorHandler.php b/core/lib/Drupal/Core/Form/FormErrorHandler.php
index a414439f8fd..878805ded68 100644
--- a/core/lib/Drupal/Core/Form/FormErrorHandler.php
+++ b/core/lib/Drupal/Core/Form/FormErrorHandler.php
@@ -5,6 +5,7 @@
 use Drupal\Component\Utility\NestedArray;
 use Drupal\Core\Messenger\MessengerTrait;
 use Drupal\Core\Render\Element;
+use Drupal\Core\Render\RendererInterface;
 
 /**
  * Handles form errors.
@@ -13,6 +14,23 @@ class FormErrorHandler implements FormErrorHandlerInterface {
 
   use MessengerTrait;
 
+  /**
+   * The renderer service.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * Constructs a FormErrorHandler instance.
+   *
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer service.
+   */
+  public function __construct(RendererInterface $renderer) {
+    $this->renderer = $renderer;
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -39,10 +57,26 @@ public function handleFormErrors(array &$form, FormStateInterface $form_state) {
    */
   protected function displayErrorMessages(array $form, FormStateInterface $form_state) {
     $errors = $form_state->getErrors();
-
-    // Loop through all form errors and set an error message.
-    foreach ($errors as $error) {
-      $this->messenger()->addMessage($error, 'error');
+    // Loop through all form errors and sets an ID & error messages.
+    foreach ($errors as $name => $error) {
+      $form_element = FormElementHelper::getElementByName($name, $form);
+      if (isset($form_element['#id'])) {
+        // Render the error messages as HTML.
+        $message = [
+          '#type' => 'container',
+          '#attributes' => [
+            'id' => $form_element['#id'] . '--error-message',
+          ],
+          'message' => [
+            '#markup' => $error,
+          ],
+        ];
+        $message = $this->renderer->renderInIsolation($message);
+        $this->messenger()->addMessage($message, 'error');
+      }
+      else {
+        $this->messenger()->addMessage($error, 'error');
+      }
     }
   }
 
@@ -164,6 +198,12 @@ protected function setElementErrorsFromFormState(array &$form, FormStateInterfac
 
     // Store the errors for this element on the element directly.
     $elements['#errors'] = $form_state->getError($elements);
+
+    // Add aria-describedby attribute to the form element.
+    if (($elements['#errors']) !== NULL) {
+      $elements['#attributes']['aria-errormessage'] = $elements['#id'] . '--error-message';
+      $elements['#attributes']['aria-invalid'] = TRUE;
+    }
   }
 
 }
diff --git a/core/lib/Drupal/Core/Render/Element/RenderElementBase.php b/core/lib/Drupal/Core/Render/Element/RenderElementBase.php
index f3b719a40ca..f203fd849cf 100644
--- a/core/lib/Drupal/Core/Render/Element/RenderElementBase.php
+++ b/core/lib/Drupal/Core/Render/Element/RenderElementBase.php
@@ -147,6 +147,9 @@ public static function setAttributes(&$element, $class = []) {
     if (isset($element['#parents']) && isset($element['#errors']) && !empty($element['#validated'])) {
       $element['#attributes']['class'][] = 'error';
       $element['#attributes']['aria-invalid'] = 'true';
+      if (isset($element['#id'])) {
+        $element['#attributes']['aria-describedby'] = $element['#id'] . '--error-message';
+      }
     }
   }
 
diff --git a/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php b/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php
index 2d4f07ff312..6506f776263 100644
--- a/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php
+++ b/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php
@@ -191,8 +191,11 @@ public function testErrorMessagesNotInline(): void {
         FALSE
       );
 
-    $this->renderer->expects($this->never())
-      ->method('renderInIsolation');
+    $this->renderer->expects($this->any())
+      ->method('renderInIsolation')
+      ->willReturnCallback(function ($message) {
+        return $message['message']['#markup'];
+      });
 
     $this->testForm['#disable_inline_form_errors'] = TRUE;
 
diff --git a/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php b/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php
index e319d52b117..569b5f77ebc 100644
--- a/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php
+++ b/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php
@@ -6,6 +6,7 @@
 
 use Drupal\Core\Form\FormState;
 use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\Core\Render\RendererInterface;
 use Drupal\Tests\UnitTestCase;
 
 /**
@@ -28,6 +29,13 @@ class FormErrorHandlerTest extends UnitTestCase {
    */
   protected $messenger;
 
+  /**
+   * The renderer service.
+   *
+   * @var \Drupal\Core\Render\RendererInterface|\PHPUnit\Framework\MockObject\MockObject
+   */
+  protected $renderer;
+
   /**
    * {@inheritdoc}
    */
@@ -36,7 +44,10 @@ protected function setUp(): void {
 
     $this->messenger = $this->createMock(MessengerInterface::class);
 
+    $this->renderer = $this->createMock(RendererInterface::class);
+
     $this->formErrorHandler = $this->getMockBuilder('Drupal\Core\Form\FormErrorHandler')
+      ->setConstructorArgs([$this->renderer])
       ->onlyMethods(['messenger'])
       ->getMock();
 
@@ -68,6 +79,12 @@ public function testDisplayErrorMessages(): void {
         'error',
       );
 
+    $this->renderer->expects($this->any())
+      ->method('renderPlain')
+      ->willReturnCallback(function ($message) {
+        return $message['message']['#markup'];
+      });
+
     $form = [
       '#parents' => [],
       '#array_parents' => [],
