diff --git a/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/TimestampFormatter.php b/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/TimestampFormatter.php
index 502b5e627bb1a50b39979aed41f6e3b9051e963d..2588f83db607980163459c6099df79fb1229a7d8 100644
--- a/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/TimestampFormatter.php
+++ b/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/TimestampFormatter.php
@@ -179,6 +179,7 @@ public function settingsForm(array $form, FormStateInterface $form_state) {
 
     $form['time_diff']['description'] = [
       '#type' => 'item',
+      '#input' => FALSE,
       '#title' => $this->t('Fallback configuration'),
       '#description' => $this->t('The configuration below is used as a fallback when JavaScript is not available on the page.'),
       '#states' => $states,
