diff --git a/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php b/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php
index 96f315aa5056e28751ffc01aa5522d54347a3e7e..83fc4f67430c26e5825baf5e28c0e8a09f0c30df 100644
--- a/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php
+++ b/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php
@@ -366,6 +366,7 @@ class ScribitProFormatter extends FormatterBase {
       '#tag' => 'iframe',
       '#attributes' => [
         'allow' => 'accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope',
+        'referrerpolicy' => 'strict-origin-when-cross-origin',
         'allowfullscreen' => 0,
         'scrolling' => 'no',
         'frameborder' => 0,
