diff --git a/composer.json b/composer.json
index e70329f88445fe534a63f102a204f7af0a85f203..40813f7c053c8b3973f6453aa05f5c847e25fbd3 100644
--- a/composer.json
+++ b/composer.json
@@ -1,6 +1,6 @@
 {
     "name": "drupal/scribit_pro",
-    "description": "Provides an integration for Scribit Pro",
+    "description": "Provides an integration for Scribit.pro",
     "type": "drupal-module",
     "license": "GPL-2.0-or-later",
     "authors": [
@@ -9,6 +9,12 @@
             "email": "tijmenve@gmail.com",
             "homepage": "https://www.drupal.org/u/tijmenve",
             "role": "Maintainer"
+        },
+        {
+            "name": "Ralph van den Houdt (ralphvdhoudt)",
+            "email": "ralph@vandenhoudt.nl",
+            "homepage": "https://www.drupal.org/u/ralphvdhoudt",
+            "role": "Maintainer"
         }
     ],
     "homepage": "https://www.drupal.org/project/scribit_pro",
diff --git a/config/schema/scribit_pro.schema.yml b/config/schema/scribit_pro.schema.yml
index 04818d486072530a6b40a11cff905fb3858eaae1..dc2e0615ffa3316f2823335cbdeeffca32cf159a 100644
--- a/config/schema/scribit_pro.schema.yml
+++ b/config/schema/scribit_pro.schema.yml
@@ -1,17 +1,17 @@
 scribit_pro.config:
   type: config_object
-  label: 'Scribit Pro configuration'
+  label: 'Scribit.pro configuration'
   mapping:
-    email:
+    api_token:
       type: string
-      label: 'Scribit Pro account email'
-    password:
+      label: 'Api Token'
+    scribit_id:
       type: string
-      label: 'Scribit Pro account password'
+      label: 'Scribit ID'
 
 field.formatter.settings.scribit_pro_formatter:
   type: mapping
-  label: 'Scribit Pro formatter settings'
+  label: 'Scribit.pro formatter settings'
   mapping:
     max_width:
       type: integer
@@ -22,4 +22,4 @@ field.formatter.settings.scribit_pro_formatter:
 
 field.widget.settings.scribit_pro_widget:
   type: field.widget.settings.string_textfield
-  label: 'Scribit Pro widget settings'
+  label: 'Scribit.pro widget settings'
diff --git a/js/scribit_pro.field_widget.js b/js/scribit_pro.field_widget.js
index b2f0d3259f8dd8a186589142855c898df2e84c8b..936d0fca94474ad726ebc8773671be898400a37e 100644
--- a/js/scribit_pro.field_widget.js
+++ b/js/scribit_pro.field_widget.js
@@ -11,7 +11,7 @@
         var voice = $(this).find('[data-scribit-pro-field="voice"]');
 
         lang.on('change', function () {
-          if (this.value == 'nl_be') {
+          if (this.value == 'nl_be' || this.value == 'it_it' || this.value == 'pl_pl' || this.value == 'es_es') {
             voice.filter('[value="male"]').attr('disabled', true);
             voice.filter('[value="female"]').prop('checked', true);
           }
diff --git a/scribit_pro.info.yml b/scribit_pro.info.yml
index d432004fbc2e1617c6e93bca433a38d347fb04b4..708bcd5701532d6c9202f0b78d2822b6f7189f7f 100644
--- a/scribit_pro.info.yml
+++ b/scribit_pro.info.yml
@@ -1,4 +1,4 @@
-name: 'Scribit Pro'
+name: 'Scribit.pro'
 type: module
 description: 'Online video accessible to all!'
 core_version_requirement: ^8 || ^9 || ^10
diff --git a/scribit_pro.install b/scribit_pro.install
new file mode 100644
index 0000000000000000000000000000000000000000..b7b20cacfd31fd0f87113cf8ffcdded99e33c538
--- /dev/null
+++ b/scribit_pro.install
@@ -0,0 +1,19 @@
+<?php
+
+/**
+ * @file
+ * Install, update and uninstall functions for the dvg_siteimprove module.
+ */
+
+/**
+ * Implements hook_install().
+ */
+
+/**
+ * Update Scribit ID.
+ */
+function scribit_pro_update_8001() {
+  $config_factory = \Drupal::configFactory();
+  $config = $config_factory->getEditable('scribit_pro.settings');
+  $config->set('scribit_id', $config->get('user_id'))->save();
+}
diff --git a/scribit_pro.module b/scribit_pro.module
index ab420cd573eae991276a1818437608b00b054e4e..ab3d17dc3946fcd466824f820363c349812188bb 100644
--- a/scribit_pro.module
+++ b/scribit_pro.module
@@ -16,7 +16,7 @@ function scribit_pro_help($route_name, RouteMatchInterface $route_match) {
     case 'help.page.scribit_pro':
       $output = '';
       $output .= '<h3>' . t('About') . '</h3>';
-      $output .= '<p>' . t('Scribit Pro') . '</p>';
+      $output .= '<p>' . t('Scribit.pro') . '</p>';
       return $output;
 
     default:
diff --git a/scribit_pro.routing.yml b/scribit_pro.routing.yml
index ef1f42d1bd08b326851bf3abc398aaf3927093bb..e2e791e3b8fdb65061a27ce7308126b831ddc580 100644
--- a/scribit_pro.routing.yml
+++ b/scribit_pro.routing.yml
@@ -2,6 +2,16 @@ scribit_pro.config:
   path: '/admin/config/system/scribit-pro'
   defaults:
     _form: '\Drupal\scribit_pro\Form\ConfigForm'
-    _title: 'Scribit Pro Configuration'
+    _title: 'Scribit.pro Configuration'
   requirements:
     _permission: 'administer scribit pro'
+
+scribit_pro.callback:
+  path: '/scribit-pro/callback'
+  defaults:
+    _controller: 'Drupal\scribit_pro\Controller\Callback::execute'
+    _title: 'Scribit.pro callback'
+  requirements:
+    _permission: 'access content'
+  options:
+    no_cache: 'TRUE'
diff --git a/src/ApiService.php b/src/ApiService.php
index 30737cc0db2f5c508c60f6c801ba5c73ddc27250..5f1e39d0fb35fe39266da7cbd63bd487ab2ab3bc 100644
--- a/src/ApiService.php
+++ b/src/ApiService.php
@@ -8,6 +8,7 @@ use Drupal\Core\Logger\LoggerChannelFactoryInterface;
 use Drupal\Core\Messenger\MessengerInterface;
 use Drupal\Core\Session\AccountProxy;
 use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\Url;
 use Drupal\key\KeyRepositoryInterface;
 use GuzzleHttp\ClientInterface;
 
@@ -80,6 +81,8 @@ class ApiService {
    */
   protected $messenger;
 
+  protected $count = 0;
+
   /**
    * Constructs a new ApiService object.
    */
@@ -94,47 +97,24 @@ class ApiService {
   }
 
   /**
-   * Get Scribit Pro projects.
-   *
-   * @return array
-   *   Array containg projects.
-   */
-  public function getProjects() {
-    return $this->authenticatedGetRequest('/v1/projects');
-  }
-
-  /**
-   * Get project for given project id.
-   *
-   * @param string $project_id
-   *   Project id.
-   *
-   * @return array
-   *   Array containg the request response.
-   */
-  public function getProject($project_id) {
-    return $this->authenticatedGetRequest('/v1/projects/' . $project_id);
-  }
-
-  /**
-   * Get project by YouTube id.
+   * Get pro services request by video url.
    *
-   * @param string $youtube_id
-   *   YouTube video id.
-   *
-   * @return object|null
+   * @param string $platform
+   * *  The plaform.
+   * *
+   * * @param string $video_id
+   * *   The video id.
+ *
+   * @return array|null
    *   Returns object if project exists for given YouTube video id. Returns null
    *   otherwise.
    */
-  public function getProjectByYouTubeId($youtube_id) {
-    $projects = $this->getProjects();
-    if (!isset($projects['body'])) {
-      return NULL;
-    }
-    foreach ($projects['body'] as $project) {
-      foreach ($project->videos as $video) {
-        if ($video->identifier == $youtube_id) {
-          return $project;
+  public function getProServicesRequest($platform, $video_id): ?array {
+    $request = $this->authenticatedGetRequest('/v1/pro-services', ['Accept: application/json, application/problem+json']);
+    if (isset($request['body'])) {
+      foreach ($request['body'] as $project) {
+        if ($project->video->external_video_id == $video_id && $project->video->platform == $platform) {
+          return json_decode(json_encode($project), true);
         }
       }
     }
@@ -143,78 +123,63 @@ class ApiService {
   }
 
   /**
-   * Post new video (project) to Scribit Pro.
+   * Sends a POST request to the Pro Services endpoint with the provided video details and service preferences.
    *
    * @param string $video_link
-   *   Video url.
+   *   The URL of the video to be processed.
    * @param string $video_id
-   *   Video id.
-   * @param string $voice_name
-   *   Voice name.
+   *   The identifier for the video.
+   * @param string $remarks
+   *   Any additional remarks or comments regarding the video.
+   * @param array $services
+   *   The list of services to be applied to the video.
+   * @param boolean $urgent
+   *   Flag indicating whether the request is urgent.
+   * @param string $voice_gender
+   *   Preferred gender for the voiceover, if applicable.
    * @param string $voice_language
-   *   Voice language.
+   *   Preferred language for the voiceover, if applicable.
    *
    * @return array
-   *   Array containing the request response.
+   *   Returns the response from the Pro Services endpoint.
    */
-  public function postProject($video_link, $video_id, $voice_name, $voice_language) {
-    $prepared_project = [
-      'external_video_link' => $video_link,
-      'external_video_id' => $video_id,
+  public function postProServicesRequest($video_link, $video_id, $remarks, $services, $urgent, $voice_gender, $voice_language): array {
+    $prepared_video = [
+      'video' => $video_link,
+      'callback' => [
+        'url' => Url::fromRoute('scribit_pro.callback', [], ['absolute' => TRUE])->toString(),
+      ],
+      'remarks' => $remarks,
+      'services' => $services,
+      'urgent' => $urgent,
       'voice' => [
-        'name' => $voice_name,
+        'gender' => $voice_gender,
         'language' => $voice_language,
       ],
     ];
-    return $this->authenticatedPostRequest('/v1/projects', ['Content-Type: application/json'], $prepared_project);
+    return $this->authenticatedPostRequest('/v1/pro-services', ['Content-Type: application/json'], $prepared_video);
   }
 
   /**
-   * Get Scribit Pro project media links for given project id.
+   * Get Scribit Pro media links for given video id.
+   *
+   * @param string $platform
+   *  The plaform.
    *
-   * @param string $project_id
-   *   Scribit Pro project id.
+   * @param string $video_id
+   *   The video id.
    *
-   * @return object
+   * @return array|null
    *   Object containg request body.
    */
-  public function getProjectMediaLinks($project_id) {
-    $request = $this->authenticatedGetRequest('/v1/projects/' . $project_id . '/requestMediaLinks');
+  public function getMediaLinks($platform, $video_id) {
+    $video_details = [];
+    $video_details[$platform] = [$video_id];
+    $request = $this->getRequest('/v1/customers/' . $this->config->get('scribit_id') . '/media-links?' . strtolower($platform) . '=' . $video_id, ['Accept: application/json, application/problem+json']);
 
     return $request['body'];
   }
 
-  /**
-   * Check if video exists for given project id.
-   *
-   * @param string $project_id
-   *   Scribit Pro project id.
-   *
-   * @return array
-   *   Array containging request response.
-   */
-  public function videoExists($project_id) {
-    return $this->authenticatedGetRequest('/v1/projects/' . $project_id . '/videoExists');
-  }
-
-  /**
-   * Sign in.
-   *
-   * @param string $email_key
-   *   Email key.
-   * @param string $password_key
-   *   Password key.
-   *
-   * @return array
-   *   Array containing the request response.
-   */
-  private function signIn($email_key = NULL, $password_key = NULL) {
-    $headers = [];
-    $headers[] = 'Authorization: Basic ' . $this->getEncodedAuthorizationCredentials($email_key, $password_key);
-
-    return $this->postRequest('/v1/jwt/signin', $headers);
-  }
-
   /**
    * Send a post request.
    *
@@ -246,9 +211,8 @@ class ApiService {
    *   Array containing the request response.
    */
   private function authenticatedPostRequest($uri, array $headers = [], array $payload = []) {
-    $token = $this->getBearerToken();
-    $headers[] = 'X-auth: ' . rtrim($token['token']);
-
+    $token = $this->getApiToken();
+    $headers[] = 'Authorization: Bearer ' . rtrim($token);
     return $this->postRequest($uri, $headers, $payload);
   }
 
@@ -283,11 +247,11 @@ class ApiService {
    *   Array containing the request response.
    */
   private function authenticatedGetRequest($uri, array $headers = [], array $payload = []) {
-    $token = $this->getBearerToken();
-    if (!isset($token['token'])) {
+    $token = $this->getApiToken();
+    if (!isset($token)) {
       return NULL;
     }
-    $headers[] = 'X-auth: ' . rtrim($token['token']);
+    $headers[] = 'Authorization: Bearer ' . rtrim($token);
 
     return $this->getRequest($uri, $headers, $payload);
   }
@@ -324,6 +288,7 @@ class ApiService {
     $content_length =  0;
     if ($payload) {
       $curl_array[CURLOPT_POSTFIELDS] = json_encode($payload, JSON_UNESCAPED_SLASHES);
+      $this->logger->debug($this->t('Payload: @headers', ['@headers' => var_export(json_encode($payload), TRUE)]));
       $content_length = strlen(json_encode($payload, JSON_UNESCAPED_SLASHES));
     }
     $headers = array_merge(
@@ -348,9 +313,9 @@ class ApiService {
       'headers' => $response_headers,
       'body' => json_decode($body),
     ];
-    if ($status !== 200) {
+    if ($status !== 200 && $status !== 201) {
       $return['error'] = TRUE;
-      $this->setStatusError($status, $return['body']);
+      $this->setStatusError($status, $return);
     }
 
     return $return;
@@ -371,7 +336,7 @@ class ApiService {
         break;
 
       case 400:
-        $this->messenger->addWarning($this->t('Something went wrong: @error', ['@error' => $body->detail]));
+        $this->messenger->addWarning($this->t('Something went wrong: @error', ['@error' => var_export($body, TRUE)]));
         break;
 
       case 404:
@@ -379,141 +344,11 @@ class ApiService {
         break;
 
       default:
-        $this->messenger->addWarning($this->t('Something went wrong connecting to Scribit Pro.'));
+        $this->messenger->addWarning($this->t('Something went wrong connecting to Scribit Pro. @status @error',['@status' => $status, '@error' => var_export($body, TRUE)]));
         break;
     }
   }
 
-  /**
-   * Refresh Bearer token.
-   *
-   * @return string|void
-   *   Bearer token.
-   */
-  private function refreshBearerToken() {
-    $refreshed_token = $this->getRequest('/v1/jwt/refresh', ['X-auth: ' . rtrim($this->getCurrentBearerToken()['token'])]);
-
-    if (!isset($refreshed_token['error'])) {
-      $new_token = $this->getBearerTokenfromHeaders($refreshed_token['headers']);
-      return $this->setBearerToken($new_token);
-    }
-  }
-
-  /**
-   * Get Bearer token.
-   *
-   * If current token is set and not expired; return current token.
-   * If the current token is expired; refresh token.
-   * If no current token is set; get a new token.
-   *
-   * @return array
-   *   Array containing token and expiration.
-   */
-  private function getBearerToken() {
-    $current_token = $this->getCurrentBearerToken();
-
-    // A token has been set and has not expired.
-    if ($current_token && isset($current_token['token']) && $current_token['token'] && isset($current_token['expires']) && time() < $current_token['expires']) {
-
-      // Token is about to expire.
-      if (time() > $current_token['expires'] - $this::EXPIRATION_OFFSET) {
-        return $this->refreshBearerToken();
-      }
-      // Token is not expired.
-      return $current_token;
-    }
-
-    // We don't have a token so get a new one.
-    return $this->getNewBearerToken();
-  }
-
-  /**
-   * Get stored Bearer token.
-   *
-   * @return array
-   *   Array containing the current token and expiration date.
-   */
-  private function getCurrentBearerToken() {
-    return $this->keyValue->get('bearer');
-  }
-
-  /**
-   * Sign in and get new Bearer token.
-   *
-   * @return array|null
-   *   Array containing the token and expiration date.
-   */
-  private function getNewBearerToken($email_key = NULL, $password_key = NULL) {
-
-    $sign_in = $this->signIn($email_key, $password_key);
-
-    if (is_array($sign_in) && is_object($sign_in['body']) && isset($sign_in['body']->user) && is_object($sign_in['body']->user) && $user = $sign_in['body']->user) {
-      $this->keyValue->set('scribit_pro_user', json_encode($user));
-      $token = $this->getBearerTokenfromHeaders($sign_in['headers']);
-
-      return [
-        'token' => $token,
-        'user' => $sign_in['body']->user,
-      ];
-    }
-
-    return NULL;
-
-  }
-
-  /**
-   * Set Bearer Token.
-   *
-   * Set and store token and expiration date.
-   *
-   * @param string $token
-   *   String containing the encoded Bearer token.
-   *
-   * @return array
-   *   Array containing the token and expiration date.
-   */
-  private function setBearerToken($token) {
-
-    // Get token parts.
-    $token_parts = explode('.', $token);
-
-    // Replace characters.
-    $data = str_replace('-', '+', $token_parts[1]);
-    $data = str_replace('_', '/', $data);
-
-    // Base64 decode.
-    $base64_decoded = base64_decode($data);
-
-    // Json decode.
-    $json_decoded = json_decode($base64_decoded);
-
-    $new_token = [
-      'token' => $token,
-      'expires' => $json_decoded->exp,
-    ];
-
-    $this->keyValue->set('bearer', $new_token);
-
-    return $new_token;
-  }
-
-  /**
-   * Extract Bearer token from returned headers.
-   *
-   * @param array $headers
-   *   Array containing response headers.
-   *
-   * @return string
-   *   Bearer token.
-   */
-  private function getBearerTokenfromHeaders(array $headers) {
-    foreach ($headers as $header) {
-      if (strtolower(substr($header, 0, 14)) == 'authorization:') {
-        return substr($header, 15);
-      }
-    }
-  }
-
   /**
    * Check if Scribit Pro credentials have been set.
    *
@@ -522,13 +357,9 @@ class ApiService {
    */
   public function hasCredentials() {
     try {
-      $email_key = $this->config->get('email');
-      $password_key = $this->config->get('password');
-
-      $email = $this->keyRepository->getKey($email_key);
-      $password = $this->keyRepository->getKey($password_key);
-
-      return ($email && $password) ? TRUE : FALSE;
+      $api_token_key = $this->config->get('api_token');
+      $api_token = $this->keyRepository->getKey($api_token_key);
+      return (bool)$api_token;
     }
     catch (\Exception $e) {
       $this->logger->warning($this->t('Something went wrong: @error', ['@error' => $e->getMessage()]));
@@ -539,61 +370,35 @@ class ApiService {
   }
 
   /**
-   * Get saved authorization credentials.
+   * Get the API Token.
    *
    * @return string
-   *   Encoded authorization credentials.
+   *   The API token.
    */
-  private function getEncodedAuthorizationCredentials($email_key = NULL, $password_key = NULL) {
-    $email_key = $email_key ? $email_key : $this->config->get('email');
-    $password_key = $password_key ? $password_key : $this->config->get('password');
-
-    $email = $this->keyRepository->getKey($email_key)->getKeyValue();
-    $password = $this->keyRepository->getKey($password_key)->getKeyValue();
-
-    return base64_encode($email . ':' . $password);
-  }
+  public function getApiToken() {
+    try {
+      $api_token_key = $this->config->get('api_token');
+      $api_token = $this->keyRepository->getKey($api_token_key);
+      return $api_token->getKeyValue();
+    }
+    catch (\Exception $e) {
+      $this->logger->warning($this->t('Something went wrong: @error', ['@error' => $e->getMessage()]));
 
-  /**
-   * Get local Scribit Pro user.
-   *
-   * Get locally stored Scribit Pro user object.
-   *
-   * @return object
-   *   Object containing the Scribit Pro user.
-   */
-  public function getLocalScribitProUser() {
-    if ($scribit_pro_user = $this->keyValue->get('scribit_pro_user')) {
-      return json_decode($scribit_pro_user);
+      return FALSE;
     }
+
   }
 
   /**
-   * Get local Scribit Pro user id.
+   * Get Scribit ID.
    *
-   * Get locally stored Scribit Pro user id.
+   * Get locally stored Scribit ID.
    *
    * @return string
-   *   Returns the locally stored Scribit Pro user id.
+   *   Returns the locally stored Scribit ID.
    */
-  public function getLocalScribitProUserId() {
-    $user = $this->getLocalScribitProUser();
-
-    return $user->id;
-  }
-
-  /**
-   * Test user login.
-   *
-   * @return object|void
-   *   User object if login succeeded.
-   */
-  public function testUserLogin($email_key, $password_key) {
-    if ($token = $this->getNewBearerToken($email_key, $password_key)) {
-      $this->setBearerToken($token['token']);
-
-      return $token['user'];
-    }
+  public function getScribitId() {
+    return $this->config->get('scribit_id');
   }
 
 }
diff --git a/src/Controller/Callback.php b/src/Controller/Callback.php
new file mode 100644
index 0000000000000000000000000000000000000000..ed2d4c0fd3c4b0b97fa9e84b684f2fb2584e5d4a
--- /dev/null
+++ b/src/Controller/Callback.php
@@ -0,0 +1,51 @@
+<?php
+
+namespace Drupal\scribit_pro\Controller;
+
+/**
+ * @file
+ * Contains \Drupal\scribit_pro\Controller\Callback.
+ */
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Messenger\MessengerTrait;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Drupal\Core\Url;
+use Drupal\Core\Render\RenderContext;
+
+/**
+ * Handles the "callback" route.
+ */
+class Callback extends ControllerBase {
+
+  use MessengerTrait;
+
+  /**
+   * Scribit.pro is calls this callback after the video is processed.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request.
+   *
+   * @return \Symfony\Component\HttpFoundation\Response
+   *   The response.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public static function execute(Request $request) {
+    $context = new RenderContext();
+    $response = \Drupal::service('renderer')->executeInRenderContext($context, function () use ($request) {
+      // TODO: handle the callback.
+      $response = new RedirectResponse(Url::fromRoute('<front>')->toString());
+      return $response;
+    });
+    // If there is metadata left on the context, apply it on the response.
+    if (!$context->isEmpty()) {
+      $metadata = $context->pop();
+      $response->addCacheableDependency($metadata);
+    }
+    return $response;
+  }
+
+}
diff --git a/src/Form/ConfigForm.php b/src/Form/ConfigForm.php
index 392fa467a088e28510911bcfbfff03c50b111f6b..f7b0adcc91e117839913edaacfff2a1bd916ddb7 100644
--- a/src/Form/ConfigForm.php
+++ b/src/Form/ConfigForm.php
@@ -77,52 +77,29 @@ class ConfigForm extends ConfigFormBase {
     $config = $this->config('scribit_pro.config');
 
     if ($input = $form_state->getUserInput()) {
-      $email = isset($input['email']) ? $input['email'] : $config->get('email');
-      $password = isset($input['password']) ? $input['password'] : $config->get('password');
+      $api_token = isset($input['api_token']) ? $input['api_token'] : $config->get('api_token');
+      $scribit_id = isset($input['scribit_id']) ? $input['scribit_id'] : $config->get('scribit_id');
     }
     else {
-      $email = $config->get('email');
-      $password = $config->get('password');
+      $api_token = $config->get('api_token');
+      $scribit_id = $config->get('scribit_id');
     }
 
-    if ($email && $password && $user = $this->helper->getScribitProUserLoginStatus($email, $password)) {
-      $form['user_status'] = [
-        '#type' => 'table',
-        '#header' => [
-          'username' => $this->t('Username'),
-          'email' => $this->t('Email'),
-        ],
-        '#rows' => [
-          [
-            $user->username,
-            $user->email,
-          ],
-        ],
-      ];
-    }
-    else {
-      $form['login_message'] = [
-        '#type' => 'item',
-        '#markup' => '<p>' . $this->t('Login using your Scribit Pro credentials.') . '</p>',
-        '#weight' => -1,
-      ];
-    }
-
-    $form['email'] = [
-      '#type' => 'key_select',
-      '#title' => $this->t('Email'),
-      '#description' => $this->t('Email'),
+    $form['scribit_id'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Scribit ID'),
+      '#description' => $this->t('The Scribit ID, can be found in your Scribit.pro account settings.'),
       '#weight' => '0',
-      '#default_value' => $email ? $email : NULL,
+      '#default_value' => $scribit_id ?? NULL,
       '#required' => TRUE,
     ];
 
-    $form['password'] = [
+    $form['api_token'] = [
       '#type' => 'key_select',
-      '#title' => $this->t('Password'),
-      '#description' => $this->t('Password'),
-      '#weight' => '1',
-      '#default_value' => $password ? $password : NULL,
+      '#title' => $this->t('Api Token'),
+      '#description' => $this->t('The Api Token, can be generated in your Scribit.pro account settings.'),
+      '#weight' => '2',
+      '#default_value' => $api_token ?? NULL,
       '#required' => TRUE,
     ];
 
@@ -144,8 +121,8 @@ class ConfigForm extends ConfigFormBase {
     $values = $form_state->getValues();
 
     $this->config('scribit_pro.config')
-      ->set('email', $values['email'])
-      ->set('password', $values['password'])
+      ->set('api_token', $values['api_token'])
+      ->set('scribit_id', $values['scribit_id'])
       ->save();
   }
 
diff --git a/src/HelperService.php b/src/HelperService.php
index c85996516df2c4d3c39bd68ff981aed7a39f036b..d1852177297e5969d5c1348e2da2530511eb8639 100644
--- a/src/HelperService.php
+++ b/src/HelperService.php
@@ -18,6 +18,11 @@ class HelperService {
    */
   private const YOUTUBE_VIDEO_URL = 'https://www.youtube.com/watch?v=';
 
+  /**
+   * Vimeo video url to be used for posting video's to Scribit Pro.
+   */
+  private const VIMEO_VIDEO_URL = 'https://vimeo.com/';
+
   /**
    * Drupal\scribit_pro\ApiService definition.
    *
@@ -33,76 +38,70 @@ class HelperService {
   }
 
   /**
-   * Get local Scribit Pro user.
-   *
-   * Get locally stored Scribit Pro user object.
-   *
-   * @return object
-   *   Object containing the Scribit Pro user.
-   */
-  public function getLocalScribitProUser() {
-    return $this->api->getLocalScribitProUser();
-  }
-
-  /**
-   * Get local Scribit Pro user id.
+   * Get local Scribit ID.
    *
-   * Get locally stored Scribit Pro user id.
+   * Get locally stored Scribit ID.
    *
    * @return string
-   *   Returns the locally stored Scribit Pro user id.
+   *   Returns the locally stored Scribit ID.
    */
-  public function getLocalScribitProUserId() {
-    return $this->api->getLocalScribitProUserId();
+  public function getScribitId() {
+    return $this->api->getScribitId();
   }
 
   /**
-   * Post youtube video to Scribit Pro.
+   * Submits a professional services request for a given video.
    *
    * @param string $video_url
-   *   YouTube video url.
+   *   URL of the video to be processed.
+   * @param string
+   *   $remarks Additional remarks or comments for the request.
+   * @param array $services
+   *   List of requested services.
+   * @param bool $urgent
+   *   Flag to indicate if the request is urgent.
    * @param string $voice_gender
-   *   Voice gender (male or female).
+   *   Preferred gender for voice-over.
    * @param string $voice_language
-   *   Scribit Pro voice language.
+   *   Preferred language for voice-over.
    *
-   * @return array
-   *   Array containing response from post request.
+   * @return mixed
+   *   The response from the API after submitting the request.
    */
-  public function postYoutubeVideo($video_url, $voice_gender, $voice_language) {
+  public function postProServicesRequest($video_url, $remarks, $services, $urgent, $voice_gender, $voice_language)
+  {
     if (!empty($video_url) && $video_id = $this->extractYouTubeIdFromUrl($video_url)) {
-      return $this->api->postProject($this->prepareYouTubeVideoUrl($video_id), $video_id, $this->getVoiceKey($voice_language, $voice_gender), $voice_language);
+      return $this->api->postProServicesRequest($this->prepareYouTubeVideoUrl($video_id), $video_id, $remarks, $services, $urgent, $voice_gender, $voice_language);
+    }
+    else if (!empty($video_url) && $video_id = $this->extractVimeoIdFromUrl($video_url)) {
+      return $this->api->postProServicesRequest($this->prepareVimeoVideoUrl($video_id), $video_id, $remarks, $services, $urgent, $voice_gender, $voice_language);
     }
   }
 
   /**
-   * Get Scribit Pro user logion status.
-   *
-   * Tests if the user can login with the known credentials.
+   * Prepare YouTube video url.
    *
-   * @param string $username_key
-   *   Username key.
-   * @param string $password_key
-   *   Password key.
+   * @param string $video_id
+   *   YouTube Video id.
    *
-   * @return object|void
-   *   Returns user object if login succeeded.
+   * @return string
+   *   Returns complete YouTube video url for Scribit Pro.
    */
-  public function getScribitProUserLoginStatus($username_key = NULL, $password_key = NULL) {
-    return $this->api->testUserLogin($username_key, $password_key);
+  public function prepareYouTubeVideoUrl($video_id) {
+    return self::YOUTUBE_VIDEO_URL . $video_id;
   }
 
   /**
-   * Prepare YouTube video url.
+   * Prepare Vimeo video url.
    *
    * @param string $video_id
-   *   YouTube Video id.
+   *   Vimeo Video id.
    *
    * @return string
-   *   Returns complete YouTube video url for Scribit Pro.
+   *   Returns complete Vimeo video url for Scribit Pro.
    */
-  public function prepareYouTubeVideoUrl($video_id) {
-    return self::YOUTUBE_VIDEO_URL . $video_id;
+  public function prepareVimeoVideoUrl($video_id) {
+    return self::VIMEO_VIDEO_URL . $video_id;
   }
 
   /**
@@ -116,10 +115,44 @@ class HelperService {
    */
   public function extractYouTubeIdFromUrl($url) {
     preg_match('/(?<=(?:v|i)=)[a-zA-Z0-9-]+(?=&)|(?<=(?:v|i)\/)[^&\n]+|(?<=embed\/)[^"&\n]+|(?<=(?:v|i)=)[^&\n]+|(?<=youtu.be\/)[^&\n]+/', $url, $id);
-
+    if (empty($id)) {
+      return NULL;
+    }
     return $id[0] ? $id[0] : NULL;
   }
 
+  /**
+   * Get YouTube id from video url.
+   *
+   * @param string $url
+   *   YouTube url the video id needs exctracted from.
+   *
+   * @return string|null
+   *   YouTube id.
+   */
+  public function extractVimeoIdFromUrl($url) {
+    preg_match('/(https?:\/\/)?(www\.)?(player\.)?vimeo\.com\/([a-z]*\/)*([0-9]{6,11})[?]?.*/', $url, $id);
+    if (empty($id)) {
+      return NULL;
+    }
+    return $id[5] ? $id[5] : NULL;
+  }
+
+  /**
+   * Get available services for Scribit Pro.
+   *
+   * @return array
+   *   Array containing available Scribit Pro Services.
+   */
+  public function getServices() {
+    return [
+      'audio description' => $this->t('Audio description'),
+      'subtitles' => $this->t('Subtitles'),
+      'transcript' => $this->t('Transcript'),
+      'sign language' => $this->t('Sign language'),
+    ];
+  }
+
   /**
    * Get available languages for Scribit Pro.
    *
@@ -132,6 +165,11 @@ class HelperService {
       'nl_nl' => $this->t('Dutch (NL)'),
       'nl_be' => $this->t('Flemish (BE)'),
       'sv_se' => $this->t('Swedish (SE)'),
+      'de_de' => $this->t('German (DE)'),
+      'it_it' => $this->t('Italian (IT)'),
+      'pl_pl' => $this->t('Polish (PL)'),
+      'fr_fr' => $this->t('French (FR)'),
+      'es_es' => $this->t('Spanish (ES)'),
     ];
   }
 
@@ -159,6 +197,23 @@ class HelperService {
         'male' => 'Sven',
         'female' => 'Karin',
       ],
+      'de_de' => [
+        'male' => 'Max',
+        'female' => 'Lena',
+      ],
+      'it_it' => [
+        'female' => 'Gina',
+      ],
+      'pl_pl' => [
+        'female' => 'Ania',
+      ],
+      'fr_fr' => [
+        'male' => 'Benoit',
+        'female' => 'Elise',
+      ],
+      'es_es' => [
+        'female' => 'Lola',
+      ],
     ];
   }
 
@@ -190,19 +245,47 @@ class HelperService {
   }
 
   /**
-   * Get project for given YouTube video url.
+   * Get pro services request for given video url.
    *
-   * @param string $youtube_url
-   *   YouTube video url.
+   * @param string $video_url
+   *   Video url.
    *
-   * @return object|null
-   *   Returns project if found, otherwise null.
+   * @return string|null
+   *   Returns pro services request id if found, otherwise null.
    */
-  public function getProjectByYouTubeUrl($youtube_url) {
-    if (!empty($youtube_url) && $youtube_id = $this->extractYouTubeIdFromUrl($youtube_url)) {
-      return $this->api->getProjectByYouTubeId($youtube_id);
+  public function getProServicesRequest($video_url) {
+    if (empty($video_url)) {
+      return NULL;
+    }
+    if ($video_id = $this->extractYouTubeIdFromUrl($video_url)) {
+      return $this->api->getProServicesRequest('YouTube', $video_id);
     }
+    else if ($video_id = $this->extractVimeoIdFromUrl($video_url)) {
+      return $this->api->getProServicesRequest('Vimeo', $video_id);
+    }
+
+    return NULL;
+  }
 
+  /**
+   * Get media links for given video url.
+   *
+   * @param string $video_url
+   *   Video url.
+   *
+   * @return array|null
+   *   Returns media links if found, otherwise null.
+   */
+  public function getMediaLinks($video_url) {
+    if (empty($video_url)) {
+      return NULL;
+    }
+    if ($video_id = $this->extractYouTubeIdFromUrl($video_url)) {
+      return $this->api->getMediaLinks('YouTube', $video_id);
+    }
+    else if ($video_id = $this->extractVimeoIdFromUrl($video_url)) {
+      return $this->api->getMediaLinks('Vimeo', $video_id);
+    }
     return NULL;
   }
 
diff --git a/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php b/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php
index 29cb3c4996e573519971768a609a26da72cf2e7f..83fc4f67430c26e5825baf5e28c0e8a09f0c30df 100644
--- a/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php
+++ b/src/Plugin/Field/FieldFormatter/ScribitProFormatter.php
@@ -356,6 +356,7 @@ class ScribitProFormatter extends FormatterBase {
     $query = [
       'enablejsapi' => 1,
       'playsinline' => 1,
+      'disablekb' => 1,
     ];
 
     $src = Url::fromUri(self::YOUTUBE_EMBED_URL . $youtube_id, ['query' => $query]);
@@ -365,6 +366,7 @@ class ScribitProFormatter extends FormatterBase {
       '#tag' => 'iframe',
       '#attributes' => [
         'allow' => 'accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope',
+        'referrerpolicy' => 'strict-origin-when-cross-origin',
         'allowfullscreen' => 0,
         'scrolling' => 'no',
         'frameborder' => 0,
@@ -378,7 +380,7 @@ class ScribitProFormatter extends FormatterBase {
     ];
 
     $embed['#attached']['library'][] = 'scribit_pro/embed';
-    $embed['#attached']['drupalSettings']['scribitPro']['user_id'] = $this->helper->getLocalScribitProUserId();
+    $embed['#attached']['drupalSettings']['scribitPro']['user_id'] = $this->helper->getScribitId();
 
     return $embed;
   }
diff --git a/src/Plugin/Field/FieldWidget/ScribitProWidget.php b/src/Plugin/Field/FieldWidget/ScribitProWidget.php
index 5ae994a314edb30a9522a4c56f827eff4ecfa538..2282358f6d9e8e8ff5a73343941bd1f9961cf90f 100644
--- a/src/Plugin/Field/FieldWidget/ScribitProWidget.php
+++ b/src/Plugin/Field/FieldWidget/ScribitProWidget.php
@@ -84,7 +84,33 @@ class ScribitProWidget extends StringTextfieldWidget {
 
     $element['scribit_pro']['#attached']['library'][] = 'scribit_pro/field_widget';
 
-    if (isset($element['value']['#default_value']) && $project = $this->helper->getProjectByYouTubeUrl($element['value']['#default_value'])) {
+    if (isset($element['value']['#default_value']) && $pro_services_request = $this->helper->getProServicesRequest($element['value']['#default_value'])) {
+      $element['scribit_pro']['video_exists_message'] = [
+        '#type' => 'item',
+        '#markup' => $this->t('This video has already been submitted to Scribit Pro'),
+      ];
+      $element['scribit_pro']['details'] = [
+        '#type' => 'fieldset',
+        '#title' => $this->t('Details'),
+        '#collapsible' => TRUE,
+        '#collapsed' => TRUE,
+      ];
+      $translated_services = [];
+      foreach ($pro_services_request['services'] as $service) {
+        $translated_services[] = $this->helper->getServices()[$service] ?? $this->t('Unknown service');
+      }
+      $report = [
+        'status' => [ 'label' => $this->t('Status'), 'value' => $pro_services_request['status'] ],
+        'services' => [ 'label' => $this->t('Services'), 'value' => implode(', ', $translated_services) ],
+        'urgent' => [ 'label' => $this->t('Urgent'), 'value' => $pro_services_request['urgent'] ? $this->t('Yes') : $this->t('No') ],
+        'remarks' => [ 'label' => $this->t('Remarks'), 'value' => $pro_services_request['remarks'] ],
+        'estimated_delivery_date' => [ 'label' => $this->t('Estimated delivery date'), 'value' => $pro_services_request['estimated_delivery_date'] ?? $this->t('Unknown') ],
+      ];
+      foreach ($report as $key => $value) {
+        $element['scribit_pro']['details'][$key] = $this->createSystemReportRow($value['label'], $value['value']);
+      }
+    }
+    elseif (isset($element['value']['#default_value']) && $this->helper->getMediaLinks($element['value']['#default_value'])) {
       $element['scribit_pro']['video_exists_message'] = [
         '#type' => 'item',
         '#markup' => $this->t('This video has already been submitted to Scribit Pro'),
@@ -100,6 +126,25 @@ class ScribitProWidget extends StringTextfieldWidget {
         ],
         '#required' => FALSE,
       ];
+      $element['scribit_pro']['services'] = [
+        '#type' => 'checkboxes',
+        '#tree' => TRUE,
+        '#title' => $this->t('Services'),
+        '#options' => $this->helper->getServices(),
+        '#attributes' => [
+          'data-scribit-pro-field' => 'services',
+        ],
+        '#states' => [
+          'visible' => [
+            ':input[data-scribit-pro-enable="1"]' => ['checked' => TRUE],
+          ],
+        ],
+        '#element_validate' => [
+          [$this, 'settingsFormRequiredValidate'],
+        ],
+        '#required' => FALSE,
+        '#description' => $this->t('Select the services you want to be added to the video.'),
+      ];
 
       $element['scribit_pro']['voice'] = [
         '#type' => 'radios',
@@ -121,6 +166,7 @@ class ScribitProWidget extends StringTextfieldWidget {
           [$this, 'settingsFormRequiredValidate'],
         ],
         '#required' => FALSE,
+        '#description' => $this->t('Select the voice you want to be used for the audio description.'),
       ];
 
       $element['scribit_pro']['language'] = [
@@ -139,6 +185,33 @@ class ScribitProWidget extends StringTextfieldWidget {
         '#element_validate' => [
           [$this, 'settingsFormRequiredValidate'],
         ],
+        '#description' => $this->t('Select the language you want to be used for the audio description.'),
+      ];
+      $element['scribit_pro']['urgent'] = [
+        '#type' => 'checkbox',
+        '#tree' => TRUE,
+        '#title' => $this->t('Urgent'),
+        '#attributes' => [
+          'data-scribit-pro-field' => 'urgent',
+        ],
+        '#states' => [
+          'visible' => [
+            ':input[data-scribit-pro-enable="1"]' => ['checked' => TRUE],
+          ],
+        ],
+        '#required' => FALSE,
+        '#description' => $this->t('This will prioritize the processing of your video. An extra charge may apply.'),
+      ];
+      $element['scribit_pro']['remarks'] = [
+        '#type' => 'textfield',
+        '#title' => $this->t('Remarks'),
+        '#description' => $this->t('Add any remarks you want to share with the Scribit Pro team.'),
+        '#states' => [
+          'visible' => [
+            ':input[data-scribit-pro-enable="1"]' => ['checked' => TRUE],
+          ],
+        ],
+        '#required' => FALSE,
       ];
     }
 
@@ -148,8 +221,8 @@ class ScribitProWidget extends StringTextfieldWidget {
 
     $description[] = $this->t('You can link to media from the following services: @providers', ['@providers' => implode(', ', $providers)]);
 
-    if (count($providers) > 1 || !in_array('YouTube', $providers)) {
-      $description[] = $this->t('Be aware that at the moment only YouTube videos can be submitted to Scribit Pro.');
+    if (!empty(array_diff($providers, ['YouTube', 'Vimeo']))) {
+      $description[] = $this->t('Be aware that at the moment only YouTube and Vimeo videos can be submitted to Scribit Pro.');
     }
 
     if (!empty($element['#value']['#description'])) {
@@ -164,6 +237,48 @@ class ScribitProWidget extends StringTextfieldWidget {
     return $element;
   }
 
+  public function createSystemReportRow($label, $value) {
+    return [
+      '#type' => 'container',
+      '#attributes' => ['class' => ['system-status-report__row']],
+      'label' => [
+        '#type' => 'html_tag',
+        '#tag' => 'div',
+        '#value' => $label,
+        '#attributes' => ['class' => ['system-status-report__status-title']],
+      ],
+      'content' => [
+        '#type' => 'html_tag',
+        '#tag' => 'div',
+        '#value' => $value,
+        '#attributes' => ['class' => ['system-status-report__entry__value']],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateElements(array &$form, FormStateInterface $form_state) {
+    parent::validateElements($form, $form_state);
+
+    $plugin_id = $this->getPluginId();
+    $assurance_levels_enabled = $form_state->getValue([
+      $plugin_id,
+      'assurance_levels_enabled',
+    ]);
+    // If the authentication provider is enabled,
+    // assurance_levels_enabled is required!
+    if ($form_state->isValueEmpty([$plugin_id, 'enabled']) === FALSE) {
+      $checked = array_filter($assurance_levels_enabled);
+      if (empty($checked)) {
+        $form_state->setError($form[$plugin_id]['assurance_levels_enabled'], $this->t('@name field is required.', [
+          '@name' => $form[$plugin_id]['assurance_levels_enabled']['#title'],
+        ]));
+      }
+    }
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -173,6 +288,9 @@ class ScribitProWidget extends StringTextfieldWidget {
     $scribit_pro_enabled[] = 'enable';
     $scribit_pro_enabled_value = $form_state->getValue($scribit_pro_enabled);
     $submitted_value = $form_state->getValue($element['#parents']);
+    if ($element['#type'] == 'checkboxes') {
+      $submitted_value = array_filter($submitted_value);
+    }
     if ($scribit_pro_enabled_value && empty($submitted_value)) {
       $form_state->setError($element, $this->t('@name field is required.', ['@name' => $element['#title']]));
     }
@@ -182,29 +300,26 @@ class ScribitProWidget extends StringTextfieldWidget {
    * {@inheritdoc}
    */
   public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
-    $prepared_values = [];
-
-    foreach ($values as $value_key => $item) {
-      foreach ($item as $item_key => $item_value) {
-        if ($item_key == 'scribit_pro') {
-          if (isset($item_value['enable']) && $item_value['enable'] == 1 && !empty($item_value['voice']) && !empty($item_value['language']) && !$this->helper->getProjectByYouTubeUrl($item['value'])) {
-            $post_video = $this->helper->postYoutubeVideo($item['value'], $item_value['voice'], $item_value['language']);
-
-            if (isset($post_video['error']) && $post_video['error']) {
-              \Drupal::messenger()->addWarning($post_video['body']->detail);
-            }
-            else {
-              \Drupal::messenger()->addStatus('Video has been submitted to Scribit Pro');
+    // Check if the form has been submitted.
+    if ($form_state->isValidationComplete()) {
+      $prepared_values = [];
+      foreach ($values as $value_key => $item) {
+        foreach ($item as $item_key => $item_value) {
+          if ($item_key == 'scribit_pro') {
+            if (isset($item_value['enable']) && $item_value['enable'] == 1 && !empty($item_value['voice']) && !empty($item_value['language']) && !empty(array_filter($item_value['services'])) && !$this->helper->getProServicesRequest($item['value'])) {
+              $post_video = $this->helper->postProServicesRequest($item['value'], $item_value['remarks'] ?? '', array_keys(array_filter($item_value['services'])), (bool)$item_value['urgent'] ?? FALSE, $item_value['voice'], $item_value['language']);
+              if (!isset($post_video['error']) || !$post_video['error']) {
+                \Drupal::messenger()->addStatus('Video has been submitted to Scribit Pro');
+              }
             }
+          } else {
+            $prepared_values[$value_key][$item_key] = $item_value;
           }
         }
-        else {
-          $prepared_values[$value_key][$item_key] = $item_value;
-        }
       }
+      return parent::massageFormValues($prepared_values, $form, $form_state);
     }
-
-    return parent::massageFormValues($prepared_values, $form, $form_state);
+    return parent::massageFormValues($values, $form, $form_state);
   }
 
 }
