diff --git a/includes/webform.theme.inc b/includes/webform.theme.inc
index 714c2b93c..f461d6dac 100644
--- a/includes/webform.theme.inc
+++ b/includes/webform.theme.inc
@@ -657,7 +657,35 @@ function webform_preprocess_file_upload_help(&$variables) {
   $upload_validators = $variables['upload_validators'];
 
   if (isset($upload_validators['webform_file_limit']) && $upload_validators['webform_file_limit'][0]) {
-    $variables['descriptions'][] = t('@size limit per form.', ['@size' => format_size($upload_validators['webform_file_limit'][0])]);
+    $new_descriptions = [];
+    $single_max_file_size = 0;
+    $form_max_file_size = $upload_validators['webform_file_limit'][0];
+    if (isset($upload_validators['file_validate_size'])) {
+      @trigger_error('\'file_validate_size\' is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the \'FileSizeLimit\' constraint instead. See https://www.drupal.org/node/3363700', E_USER_DEPRECATED);
+      $single_max_file_size = $upload_validators['file_validate_size'][0];
+    }
+    if (isset($upload_validators['FileSizeLimit'])) {
+      $single_max_file_size = $upload_validators['FileSizeLimit']['fileLimit'];
+    }
+    foreach ($variables['descriptions'] as $description) {
+      if ($description->getUntranslatedString() !== '@size limit.') {
+        $new_descriptions[] = $description;
+      }
+    }
+    if ($single_max_file_size < $form_max_file_size) {
+      $new_descriptions[] = t('@file_size limit per file. The accumulated size of all files in this form cannot exceed @size.',
+        [
+          '@file_size' => format_size($single_max_file_size),
+          '@size' => format_size($form_max_file_size),
+        ]);
+    }
+    else {
+      $new_descriptions[] = t('The accumulated size of all files in this form cannot exceed @size.',
+        [
+          '@size' => format_size($form_max_file_size),
+        ]);
+    }
+    $variables['descriptions'] = $new_descriptions;
   }
 
   // Issue #3136578: Comma-separate the list of allowed file extensions.
diff --git a/tests/src/Functional/Element/WebformElementManagedFileLimitTest.php b/tests/src/Functional/Element/WebformElementManagedFileLimitTest.php
index 8d23632a5..67416668e 100644
--- a/tests/src/Functional/Element/WebformElementManagedFileLimitTest.php
+++ b/tests/src/Functional/Element/WebformElementManagedFileLimitTest.php
@@ -43,12 +43,12 @@ class WebformElementManagedFileLimitTest extends WebformElementManagedFileTestBa
 
     // Check form file limit.
     $this->drupalGet('/webform/test_element_managed_file_limit');
-    $assert_session->responseContains('1 MB limit per form.');
+    $assert_session->responseContains('The accumulated size of all files in this form cannot exceed 1 MB.');
 
     // Check empty form file limit.
     $webform->setSetting('form_file_limit', '')->save();
     $this->drupalGet('/webform/test_element_managed_file_limit');
-    $assert_session->responseNotContains('1 MB limit per form.');
+    $assert_session->responseNotContains('The accumulated size of all files in this form cannot exceed 1 MB.');
 
     // Check default form file limit.
     \Drupal::configFactory()
@@ -56,7 +56,7 @@ class WebformElementManagedFileLimitTest extends WebformElementManagedFileTestBa
       ->set('settings.default_form_file_limit', '2 MB')
       ->save();
     $this->drupalGet('/webform/test_element_managed_file_limit');
-    $assert_session->responseContains('2 MB limit per form.');
+    $assert_session->responseContains('The accumulated size of all files in this form cannot exceed 2 MB.');
 
     // Set limit to 2 files.
     \Drupal::configFactory()
@@ -64,7 +64,7 @@ class WebformElementManagedFileLimitTest extends WebformElementManagedFileTestBa
       ->set('settings.default_form_file_limit', ($bytes * 2) . ' bytes')
       ->save();
     $this->drupalGet('/webform/test_element_managed_file_limit');
-    $assert_session->responseContains(format_size($bytes * 2) . ' limit per form.');
+    $assert_session->responseContains('The accumulated size of all files in this form cannot exceed ' . format_size($bytes * 2));
 
     // Check valid file upload.
     $edit = [
