diff --git a/.cspell-project-words.txt b/.cspell-project-words.txt
index 1f34c35e83dc2c145548dd159b02ee85d041f22f..a32b5e305a77c61509387fc3798a284a242921e0 100644
--- a/.cspell-project-words.txt
+++ b/.cspell-project-words.txt
@@ -15,7 +15,9 @@ bestuurscolleges
 bereikbaarheidsgegevens
 bijlagen
 categorie
+convenant
 convenanten
+convenants
 decentrale
 depublish
 facetable
@@ -48,6 +50,7 @@ overheid
 overige
 oinregister
 publicatiedatum
+publicatie
 publicaties
 regelgeving
 resup
diff --git a/modules/openwoo_search/js/openwoo_search.js b/modules/openwoo_search/js/openwoo_search.js
index 2305878e5c2d6e2a65281dfc03c370fa492d2d38..f32f8b0585e8a617cf562dc9bf0227b65cd36aa0 100644
--- a/modules/openwoo_search/js/openwoo_search.js
+++ b/modules/openwoo_search/js/openwoo_search.js
@@ -11,7 +11,7 @@
           ).value;
           const yearValue = document.querySelector('select[name="year"]').value;
           const categoryValue = document.querySelector(
-            'select[name="category_id"]',
+            '[name="category_id"]',
           ).value;
           const url = document.querySelector(
             'input[name="ajax_filter_url"]',
diff --git a/modules/openwoo_search/src/Form/OpenWooAjaxForm.php b/modules/openwoo_search/src/Form/OpenWooAjaxForm.php
index c827652cca410f4125c3f2d00c29b5c92ddcde0b..dee6a5e55a2cf482a4288d73d82e1259c9979109 100644
--- a/modules/openwoo_search/src/Form/OpenWooAjaxForm.php
+++ b/modules/openwoo_search/src/Form/OpenWooAjaxForm.php
@@ -120,9 +120,7 @@ class OpenWooAjaxForm extends FormBase implements ContainerInjectionInterface {
     $form['openwoo_wrapper'] = $this->getOpenWooWrapper();
 
     $form['openwoo_input_wrapper']['search_text'] = $this->getSearchField();
-    if (empty($default_category_id)) {
-      $form['openwoo_input_wrapper']['category_id'] = $this->getCategoryField();
-    }
+    $form['openwoo_input_wrapper']['category_id'] = $this->getCategoryField($default_category_id);
     $form['openwoo_input_wrapper']['year'] = $this->getYearField();
     $form['openwoo_input_wrapper']['submit'] = $this->getSubmitButton();
     $form['openwoo_input_wrapper']['ajax_filter_url'] = $this->getAjaxFilterUrl();
@@ -187,15 +185,21 @@ class OpenWooAjaxForm extends FormBase implements ContainerInjectionInterface {
    * @return array
    *   The category field.
    */
-  protected function getCategoryField(): array {
+  protected function getCategoryField(string $default_category_id): array {
+    if (empty($default_category_id)) {
+      return [
+        '#type' => 'select',
+        '#options' => $this->activeSearchPlugin->getCategoryOptions(),
+        '#empty_value' => NULL,
+        '#empty_option' => $this->t('- All categories -'),
+        '#attributes' => [
+          'aria-controls' => 'openwoo-wrapper',
+        ],
+      ];
+    }
     return [
-      '#type' => 'select',
-      '#options' => $this->activeSearchPlugin->getCategoryOptions(),
-      '#empty_value' => NULL,
-      '#empty_option' => $this->t('- All categories -'),
-      '#attributes' => [
-        'aria-controls' => 'openwoo-wrapper',
-      ],
+      '#type' => 'hidden',
+      '#value' => $default_category_id,
     ];
   }
 
